nexus/src/external_api/http_entrypoints.rs

```diff
@@ -2065,19 +2095,69 @@ async fn instance_create(
     let new_instance_params = &new_instance.into_inner();
     let handler = async {
         let opctx = OpContext::for_external_api(&rqctx).await?;
-        let instance = nexus
-            .project_create_instance(
+        let project_id = nexus
+            .project_lookup_id(
                 &opctx,
-                &organization_name,
-                &project_name,
-                &new_instance_params,
+                params::ProjectSelector::ProjectAndOrg {
+                    project_name: project_name.clone().into(),
+                    organization_name: organization_name.clone().into(),
+                },
             )
             .await?;
+        let instance = nexus
+            .project_create_instance(&opctx, &project_id, &new_instance_params)
+            .await?;
         Ok(HttpResponseCreated(instance.into()))
     };
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+/// Path parameters for Instance requests
+#[derive(Deserialize, JsonSchema)]
+struct InstanceLookupPathParam {
+    /// If Name is used to reference the instance you must also include one of the following qualifiers as query parameters:
+    /// - `project_id`
+    /// - `project_name`, `organization_id`
+    /// - `project_name`, `organization_name`
+    ///
+    /// If Id is used the above qualifiers are will be ignored
+    instance: NameOrId,
+}
+
+#[derive(Deserialize, JsonSchema)]
+struct InstanceQueryParams {
+    #[serde(flatten)]
+    selector: params::ProjectSelector,
+}
+
+#[endpoint {
+    method = GET,
+    path = "/v1/instances/{instance}",
+    tags = ["instances"],
+}]
+async fn v1_instance_view(
+    rqctx: Arc<RequestContext<Arc<ServerContext>>>,
+    query_params: Query<InstanceQueryParams>,
+    path_params: Path<InstanceLookupPathParam>,
+) -> Result<HttpResponseOk<Instance>, HttpError> {
+    let apictx = rqctx.context();
+    let nexus = &apictx.nexus;
+    let path = path_params.into_inner();
+    let query = query_params.into_inner();
+    let handler = async {
+        let opctx = OpContext::for_external_api(&rqctx).await?;
+        let instance_id = nexus
+            .instance_lookup_id(
+                &opctx,
+                query.selector.to_instance_selector(path.instance.into()),
+            )
+            .await?;
```

Comment:
Even though these top level instance endpoints actually take in the project selector via query params we transform it into an instance selector for more uniform matching. This does mean that the `instance_lookup_id` method isn't exactly aware of where it's being called from and is thus can only really return a generic error message. We should probably have a custom error type for this method such that we can do a better job from the callsite here of giving the user accurate error feedback (E.g. you're missing a query param).